custom_components/tuya_local/device.py

```diff
@@ -403,6 +403,8 @@ async def _retry_on_failed_connection(self, func, error_message):
                 self._api_protocol_working = True
                 return retval
             except Exception as e:
+                for entity in self._children:
```

Comment:
I think this would be better below where we call self._reset_cached_state() after reaching the retry limit, as that is what actually causes the side effect of making the devices go unavailable.  Otherwise there could be a race condition where HA updates the state before continuing on to that line.

Reply:
I thought of that at first, but I wanted the device to go unavailable since the first network error (if the error is temporary, the device would show available later) but if that leads to unexpected consequences (I trust you, you know better than me !), I will move it below, it will be a good start !

Reply:
It anyway won't go unavailable until the conditions for it to be unavailable exist, which is after the call to self._reset_cached_state()

Reply:
Well, I've got another idea, can't we reset the state on every failure ? You tell me.

```python
          except Exception as e:
                self._reset_cached_state()
                for entity in self._children:
                    entity.async_schedule_update_ha_state()
                _LOGGER.debug(
                    f"Retrying after exception {e} ({i}/{connections})",
                )
                if i + 1 == connections:
                    self._api_protocol_working = False
                    _LOGGER.error(error_message)
                if not self._api_protocol_working:
                    await self._rotate_api_protocol_version()
```

Reply:
And yeah you(re right it was stupid at first, calling HA to make update before resetting state would not make the device unavailable since the unavailable property depends on the size of the state...

Reply:
I don't think it is a good idea, one off failures are actually quite frequent and the devices do not always return full state so it might take some time to get all the cached data back.  The retry limit is 3 once the device is communicating (and before it is communicating it is already unavailable), so it should not be too much delay.   Whatever you are seeing after your change is what you will get, just moving it avoids the possiblity of a race condition where it might take longer.

Reply:
Change is done, I let you resolve this conversation if it's ok for you.